Ability to set and reset default system prompt

humphd opened the pull request:
Fixes #138, addresses part of https://github.com/tarasglek/chatcraft.org/issues/194#issuecomment-1661036662.

When we create a new chat, we always begin with a default system prompt.  The text of this system prompt is baked into the app's code.  We allow the user to edit it in the current message, but you can't use it again in a new chat.

This change allows the default system prompt to be overridden by a user (or reset back to the default).  The process is this:

1. User edits the current system prompt
2. In the 3-dots menu for the system prompt, a new set of options is shown when the system prompt has been changed (i.e., it is NOT shown if it's not needed):
    1. Users can set the default system prompt to the current system prompt (i.e., override)
    2. Users can reset the overridden system prompt back to the default (i.e., reset)

Here's a video showing how this works in practice (sorry it's small, GitHub won't let me post anything > 10M):

https://github.com/tarasglek/chatcraft.org/assets/427398/2491af38-6bca-4c89-b981-b7fbb69f9450

@steven4354 can you try this and comment?  NOTE: I'm not tackling the part about saving multiple custom prompts and choosing between them, mainly because I'm not sure how to do the UI.  I wonder if we should start with this and iterate?  Maybe you have ideas now.

humphd commented:
For showing previous system prompts, I could easily query the db to get all previous prompts, maybe sorted by date used or something (i.e., I don't need to add any new storage for this to work).  But how do I show it?  A drop-down for something that fits in a textarea isn't really going to work (not enough context to pick them).

steven4354 commented:
Just took a look at the video, that looks great and would solve the problem for me

**A drop-down for something that fits in a textarea isn't really going to work (not enough context to pick them).**

Not sure if I'm 100% on the same track (pls lmk if I am), but are you saying the drop-down of the prompt titles? 

If so, I guess you would need a flow to "name" each system prompt as well right, maybe set a max length for the title and then only show in the dropdown the ones that they "named" + "saved"? 

So "saving" and "naming" a system prompt is required to show in the dropdown

humphd commented:
Thanks for the feedback.  Here's a concept to let the user choose from previous system prompts.  I'm borrowing UX from the way that AI Message versions work, giving the system message a dropdown in the heading:

<img width="912" alt="Screenshot 2023-08-01 at 9 14 49 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/f5057a58-a6ca-4dff-bca0-d453cc11223f">

Clicking that reveals your old system prompts, sorted by most recently used, and clicking one updates the current system prompt to use it:

<img width="985" alt="Screenshot 2023-08-01 at 9 15 23 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/498f127f-f763-44f7-912c-cdbe5a7d5962">

There's a scrolling/height bug that this exposes, but it can be fixed in a follow-up.

I also realized that I need a way to set a *different* custom system prompt in my menu solution (i.e., if you already have a custom system prompt, but you want to use a different one, we need to somehow differentiate it).

Take a look at this and see if things make sense, or what you'd do differently.

humphd commented:
I've pushed my updated code, if people want to try it out.  You might want to export/import your db from production so you can get the full effect (do that in the Settings modal).

tarasglek commented:
I like the system prompt dropdown concept. Why not add a rename button in *ChatCraft (System Prompt)*, so we could list them by name

tarasglek commented:
Options 2 would be ok for me.

humphd commented:
OK, I've reworked this to allow users to "star" a system prompt, and then only show the ones they've starred.  I also reworked the way the default system prompt works to let you do it via its own dedicated modal.  Trying to do this via the menu was too complicated, and I think this is simpler.

<img width="729" alt="Screenshot 2023-08-10 at 12 21 45 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/98f1dbd0-75c6-45b6-86d4-621400d75acb">
<img width="735" alt="Screenshot 2023-08-10 at 12 17 26 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/e20497b9-bd39-4218-81fb-b73c5d19ece8">
<img width="735" alt="Screenshot 2023-08-10 at 12 17 42 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/14d69200-5eee-48f0-ab0a-71f57189d6a7">

<img width="515" alt="Screenshot 2023-08-10 at 12 17 14 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/b2d1c210-2792-4f33-87b3-726b56c576db">

NOTE: this change needs #198 to get scrolling to not jump to the bottom when you star/unstar the system prompt.

tarasglek commented:
I just tried this.
1. edited the system prompt
2. stared it
3. now I have no more default prompt in the menu. This is scary.
<img width="946" alt="image" src="https://github.com/tarasglek/chatcraft.org/assets/857083/6667a3d5-41f8-4209-80a1-a7700fc9ae03">

"new blank chat" brings the old prompt back, but menu still doesn't show it. We should star the default prompt.

Feels like a scary amount of state for a very corner-case feature...How about tagging shared chats as "chat-templates".

humphd commented:
To clarify what I'm reading above, you're saying the default system prompt should also be in that drop-down menu?  If so, that's a good idea.

humphd commented:
Rebased and updated to include the default system prompt as first item in list of starred entries.

tarasglek commented:
I wonder about having this sort of complexity in indexeddb...For me it would be better to keep such a "personal prompt library" in cloud so i could use it across all my devices. Same goes for the /commands and @functions we are talking about in other issue

humphd commented:
"keep...in the cloud" is scope creep.  If we want to explore that, let's not do it in this PR.

How about the functionality of this change?  What do I need to do in order to finish it?  Same question to @steven4354.

tarasglek commented:
I def think this is sufficient. Not asking for scope creep, just noting that you are taking on a more complex client-side schema with this change.

tarasglek commented:
lets land the last version you posted

humphd commented:
Rebased, updated `toast()` calls to use new `useAlert()` hook, will land when this goes green...

humphd commented:
Go forth and create reusable system prompts!

steven4354 commented:
I'm going to use this a bunch :)
